### PR TITLE
Added monitoring to sync calls & made all sync calls the same scheme.

### LINKED ISSUE #93: available_worker doesn't handle worker exit
With strategy available_worker, at some point, call is transformed into a cast, which means there's no monitor on the worker process, and if it dies as the result, the call times out instead of failing otherwise.

### PR DESCRIPTION
**Pull request by ElectronicRU**
Now it follows the "cast a call" paradigm for all types of all calls,
and all_state_events are now queued correctly, too.

Fixes #93.

### CODE REVIEW

File: src/wpool_queue_manager.erl [hunk 1]
```diff
@@ -311,9 +360,9 @@ handle_call(
       case erlang:is_process_alive(ClientPid) andalso
         Expires > now_in_microseconds() of
         true  ->
-          Reply = wpool_fsm_process:sync_send_event(Worker, Event),
-          gen_server:reply(Client, Reply),
-          {noreply, State#state{workers = NewWorkers}};
+          NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 363 is too long:           NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),.

File: src/wpool_queue_manager.erl [hunk 2]
```diff
@@ -182,7 +200,7 @@ stats(PoolName) ->
 -spec init(wpool:name()) -> {ok, state()}.
 init(WPool) ->
   put(pending_tasks, 0),
-  {ok, #state{wpool = WPool, clients = queue:new(), workers = gb_sets:new()}}.
+  {ok, #state{wpool = WPool, clients = queue:new(), workers = gb_sets:new(), monitors = gb_trees:new()}}.
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 203 is too long:   {ok, #state{wpool = WPool, clients = queue:new(), workers = gb_sets:new(), monitors = gb_trees:new()}}..

File: src/wpool_queue_manager.erl [hunk 3]
```diff
@@ -106,14 +121,17 @@ sync_send_event_to_available_worker(QueueManager, Event, Timeout) ->
...
-    end,
+  Expires = expires(Timeout),
   try
     gen_server:call(
       QueueManager, {sync_all_event_available_worker, Event, Expires}, Timeout)
+  of
+    {'EXIT', _, noproc} ->
+      noproc;
+    {'EXIT', Worker, Exit} ->
+      exit({Exit, {gen_fsm, sync_send_all_state_event, [Worker, Event, Timeout]}});
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 132 is too long:       exit({Exit, {gen_fsm, sync_send_all_state_event, [Worker, Event, Timeout]}});.

File: src/wpool_queue_manager.erl [hunk 4]
```diff
@@ -261,7 +309,7 @@ handle_cast({send_all_event_to_available_worker, Event}, State) ->
   case gb_sets:is_empty(Workers) of
     true ->
       inc_pending_tasks(),
-      {noreply, State#state{clients = queue:in({send_event, Event}, Clients)}};
+      {noreply, State#state{clients = queue:in({send_all_event, Event}, Clients)}};
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 312 is too long:       {noreply, State#state{clients = queue:in({send_all_event, Event}, Clients)}};.

File: src/wpool_queue_manager.erl [hunk 5]
```diff
@@ -326,17 +375,18 @@ handle_call(
...
     false ->
       {Worker, NewWorkers} = gb_sets:take_smallest(Workers),
       %NOTE: It could've been a while since this call was made, so we check
       case erlang:is_process_alive(ClientPid) andalso
         Expires > now_in_microseconds() of
         true  ->
-          Reply = wpool_fsm_process:sync_send_all_state_event(Worker, Event),
-          gen_server:reply(Client, Reply),
-          {noreply, State#state{workers = NewWorkers}};
+          NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 387 is too long:           NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),.

File: src/wpool_queue_manager.erl [hunk 6]
```diff
@@ -288,8 +336,9 @@ handle_call(
       case erlang:is_process_alive(ClientPid) andalso
            Expires > now_in_microseconds() of
         true  ->
+          NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 339 is too long:           NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),.

File: src/wpool_queue_manager.erl [hunk 7]
```diff
@@ -348,6 +398,15 @@ handle_call(worker_counts, _From, State) ->
 
 %% @private
 -spec handle_info(any(), state()) -> {noreply, state()}.
+handle_info({'DOWN', _Ref, process, Worker, Exit}, State = #state{monitors = Mons}) ->
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 401 is too long: handle_info({'DOWN', _Ref, process, Worker, Exit}, State = #state{monitors = Mons}) ->.

File: src/wpool_queue_manager.erl [hunk 8]
```diff
@@ -48,6 +48,7 @@
 -record(state, { wpool                 :: wpool:name()
                , clients               :: queue:queue({cast|{pid(), _}, term()})
                , workers               :: gb_sets:set(atom())
+               , monitors              :: gb_trees:tree(atom(), {reference(), {pid(), any()}})
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 51 is too long:                , monitors              :: gb_trees:tree(atom(), {reference(), {pid(), any()}}).

File: src/wpool_fsm_process.erl [hunk 9]
```diff
@@ -177,6 +183,20 @@ format_status(Opt, [PDict, StateData]) ->
 %%%===================================================================
 -spec handle_event(term(), fsm_state(), state()) ->
   {next_state, dispatch_state, state()} | {stop, term(), state()}.
+handle_event({sync_send_event, From, Event}, StateName, StateData) ->
+  case handle_sync_event(Event, From, StateName, StateData) of
+    {reply, Reply, NextState, NextData} ->
+      gen_fsm:reply(From, Reply),
+      {next_state, NextState, NextData};
+    {reply, Reply, NextState, NextData, Timeout} ->
```
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Dialyzer**:
> The pattern {'reply', Reply, NextState, NextData, Timeout} can never match the type {'next_state','dispatch_state',#state{name::atom(),mod::atom(),options::[{'overrun_handler' | 'overrun_warning' | 'pool_sup_intensity' | 'pool_sup_period' | 'queue_manager' | 'strategy' | 'time_checker' | 'worker' | 'worker_opt' | 'worker_type' | 'workers',_}],fsm_state::atom()}} | {'stop',_,#state{name::atom(),mod::atom(),options::[{'overrun_handler' | 'overrun_warning' | 'pool_sup_intensity' | 'pool_sup_period' | 'queue_manager' | 'strategy' | 'time_checker' | 'worker' | 'worker_opt' | 'worker_type' | 'workers',_}],fsm_state::atom()}}

File: src/wpool_queue_manager.erl (hunk 2 above)
Comment (gadgetci) [inline src/wpool_queue_manager.erl]: According to **Dialyzer**:
> Call to missing or unexported function gb_trees:new/0

File: src/wpool_fsm_process.erl [hunk 10]
```diff
@@ -177,6 +183,20 @@ format_status(Opt, [PDict, StateData]) ->
...
+      gen_fsm:reply(From, Reply),
+      {next_state, NextState, NextData, Timeout};
+    {stop, Reason, Reply, NextData} ->
```
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Dialyzer**:
> The pattern {'stop', Reason, Reply, NextData} can never match the type {'next_state','dispatch_state',#state{name::atom(),mod::atom(),options::[{'overrun_handler' | 'overrun_warning' | 'pool_sup_intensity' | 'pool_sup_period' | 'queue_manager' | 'strategy' | 'time_checker' | 'worker' | 'worker_opt' | 'worker_type' | 'workers',_}],fsm_state::atom()}} | {'stop',_,#state{name::atom(),mod::atom(),options::[{'overrun_handler' | 'overrun_warning' | 'pool_sup_intensity' | 'pool_sup_period' | 'queue_manager' | 'strategy' | 'time_checker' | 'worker' | 'worker_opt' | 'worker_type' | 'workers',_}],fsm_state::atom()}}

File: src/wpool_queue_manager.erl [hunk 11]
```diff
@@ -326,17 +379,18 @@ handle_call(
...
     false ->
       {Worker, NewWorkers} = gb_sets:take_smallest(Workers),
       %NOTE: It could've been a while since this call was made, so we check
       case erlang:is_process_alive(ClientPid) andalso
         Expires > now_in_microseconds() of
         true  ->
-          Reply = wpool_fsm_process:sync_send_all_state_event(Worker, Event),
-          gen_server:reply(Client, Reply),
-          {noreply, State#state{workers = NewWorkers}};
+          NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 391 is too long:           NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),.

File: src/wpool_queue_manager.erl [hunk 12]
```diff
@@ -261,7 +311,7 @@ handle_cast({send_all_event_to_available_worker, Event}, State) ->
   case gb_sets:is_empty(Workers) of
     true ->
       inc_pending_tasks(),
-      {noreply, State#state{clients = queue:in({send_event, Event}, Clients)}};
+      {noreply, State#state{clients = queue:in({send_all_event, Event}, Clients)}};
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 314 is too long:       {noreply, State#state{clients = queue:in({send_all_event, Event}, Clients)}};.

File: src/wpool_queue_manager.erl [hunk 13]
```diff
@@ -348,6 +402,15 @@ handle_call(worker_counts, _From, State) ->
 
 %% @private
 -spec handle_info(any(), state()) -> {noreply, state()}.
+handle_info({'DOWN', _Ref, process, Worker, Exit}, State = #state{monitors = Mons}) ->
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 405 is too long: handle_info({'DOWN', _Ref, process, Worker, Exit}, State = #state{monitors = Mons}) ->.

File: src/wpool_queue_manager.erl [hunk 14]
```diff
@@ -48,6 +48,7 @@
 -record(state, { wpool                 :: wpool:name()
                , clients               :: queue:queue({cast|{pid(), _}, term()})
                , workers               :: gb_sets:set(atom())
+               , monitors              :: gb_trees:tree(atom(), {reference(), from()})
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 51 is too long:                , monitors              :: gb_trees:tree(atom(), {reference(), from()}).

**Comment by elbrujohalcon** [comment]
@ElectronicRU: @elvisci is not happy with this PR, can you satisfy him :trollface: ?

**Comment by ElectronicRU** [comment]
Dammit, what with that 80-character lines? Nobody I know uses dumb terminals anymore. :(

**Comment by elbrujohalcon** [comment]
@ElectronicRU a couple of us use split-screens at @inaka, we need 80 columns on each side to be able to read 2 files at the same time. Besides, it's in [our guidelines](http://github.com/inaka/erlang_guidelines)

File: src/wpool_queue_manager.erl [hunk 15]
```diff
@@ -348,6 +403,15 @@ handle_call(worker_counts, _From, State) ->
 
 %% @private
 -spec handle_info(any(), state()) -> {noreply, state()}.
+handle_info({'DOWN', _Ref, process, Worker, Exit}, State = #state{monitors = Mons}) ->
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 406 is too long: handle_info({'DOWN', _Ref, process, Worker, Exit}, State = #state{monitors = Mons}) ->.

File: src/wpool_queue_manager.erl [hunk 16]
```diff
@@ -261,7 +312,7 @@ handle_cast({send_all_event_to_available_worker, Event}, State) ->
   case gb_sets:is_empty(Workers) of
     true ->
       inc_pending_tasks(),
-      {noreply, State#state{clients = queue:in({send_event, Event}, Clients)}};
+      {noreply, State#state{clients = queue:in({send_all_event, Event}, Clients)}};
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 315 is too long:       {noreply, State#state{clients = queue:in({send_all_event, Event}, Clients)}};.

File: src/wpool_queue_manager.erl [hunk 17]
```diff
@@ -48,10 +48,12 @@
 -record(state, { wpool                 :: wpool:name()
                , clients               :: queue:queue({cast|{pid(), _}, term()})
                , workers               :: gb_sets:set(atom())
+               , monitors              :: gb_trees:tree(atom(), monitored_from())
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 51 is too long:                , monitors              :: gb_trees:tree(atom(), monitored_from()).

File: src/wpool_queue_manager.erl [hunk 18]
```diff
@@ -326,17 +380,18 @@ handle_call(
...
     false ->
       {Worker, NewWorkers} = gb_sets:take_smallest(Workers),
       %NOTE: It could've been a while since this call was made, so we check
       case erlang:is_process_alive(ClientPid) andalso
         Expires > now_in_microseconds() of
         true  ->
-          Reply = wpool_fsm_process:sync_send_all_state_event(Worker, Event),
-          gen_server:reply(Client, Reply),
-          {noreply, State#state{workers = NewWorkers}};
+          NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),
```
Comment (elvisci) [inline src/wpool_queue_manager.erl]: Line 392 is too long:           NewState = monitor_worker(Worker, Client, State#state{workers = NewWorkers}),.

**Comment by elbrujohalcon** [comment]
@ElectronicRU Elvis is still unhappy. You can check that locally by running the tests with `rebar ct` or running `$ elvis rock` in your command line.

**Comment by elbrujohalcon** [comment]
Sorry, @ElectronicRU, but elvis is still not pleased. I'm sure you'll see what to fix if you run `rebar3 ct`.

**Comment by elbrujohalcon** [comment]
Sorry, @ElectronicRU, but elvis is still unhappy. I'm sure you can see what he complains about by running `rebar3 ct`.

**Comment by elbrujohalcon** [comment]
@ElectronicRU I would love to merge your PR… could you please get `rebar3 ct` to pass?

**Comment by ElectronicRU** [comment]
I'm sorry, I'm just really booked right now. 'Ll try to take a look next week.

**Comment by elbrujohalcon** [comment]
@ElectronicRU don't worry… I just figured I can do the changes myself :)

**Pull request merged**